### PR TITLE
Uxprod 2627

### PR DESCRIPTION
**Pull request by LeireDiez**
## Purpose

MODCAT-210: Create authority from json 

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this introduce breaking changes?
  - [x] Were any API paths or methods changed, added or removed?
  - [x] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: ramls/examples/name.json
Comment (carmentrazza) [inline ramls/examples/name.json:1]: The structure of the tags must be the same as the bibliographic ones.
Some properties are missing for both fixed and variable tags, 
see record.json of bibliographic records in tests, example:
 {
        "code": "008",
        "mandatory": true,
        "fieldStatus": "unchanged",
        "fixedField": {
          "categoryCode": 1,
          "headerTypeCode": 31,
          "code": "008",
          "displayValue": "180924s        enk           000 u eng r",
          "sequenceNumber": 0,
          "attributes": {}
        },



 {
        "code": "100",
        "mandatory": false,
        "fieldStatus": "new",
        "variableField": {
          "keyNumber": 0,
          "code": "100",
          "ind1": "1",
          "ind2": " ",
          "displayValue": "\u001faAlessandro Manzoni,\u001fd1785-1873.",
          "headingTypeCode": 0,
          "sequenceNumber": 0,
          "skipInFiling": 0,
          "subfields": [],
          "categoryCode": 2
        },
        "added": true
      },

File: ramls/schema/authority.json
Comment (carmentrazza) [inline ramls/schema/authority.json:2]: Add missing properties for fixed and variable tags

File: src/main/java/org/folio/marccat/config/constants/Global.java
Comment (eturienzo) [reply src/main/java/org/folio/marccat/config/constants/Global.java]: I have done the name change

File: src/test/java/org/folio/marccat/resources/AuthorityRecordTest.java
Comment (eturienzo) [reply src/test/java/org/folio/marccat/resources/AuthorityRecordTest.java:26]: At the moment there is not necesary because we have only implemented name authority

File: src/main/java/org/folio/marccat/config/log/Message.java
Comment (eturienzo) [reply src/main/java/org/folio/marccat/config/log/Message.java:15]: I have changed the number

File: src/main/resources/sample-data/load-from-file/authority/name.json
Comment (eturienzo) [reply src/main/resources/sample-data/load-from-file/authority/name.json:1]: We have created the authority records from json file and not from mrc

File: src/main/java/org/folio/marccat/integration/AuthorityStorageService.java
Comment (eturienzo) [reply src/main/java/org/folio/marccat/integration/AuthorityStorageService.java:175]: I know that it is for bibliographic. Can I use for the title authority record?

File: src/main/java/org/folio/marccat/integration/AuthorityStorageService.java
Comment (eturienzo) [reply src/main/java/org/folio/marccat/integration/AuthorityStorageService.java]: I have done the change

File: src/main/java/org/folio/marccat/integration/AuthorityStorageService.java
Comment (eturienzo) [reply src/main/java/org/folio/marccat/integration/AuthorityStorageService.java:92]: At the moment we don't use templates

File: src/test/resources/marccat.sql
Comment (eturienzo) [reply src/test/resources/marccat.sql]: I have committed the previous one.

File: src/main/java/org/folio/marccat/dao/persistence/ItemEntity.java
Comment (eturienzo) [reply src/main/java/org/folio/marccat/dao/persistence/ItemEntity.java:25]: In authority MARC the  Encoding level (leader-17) has the values 'n' or 'o'. Why should it be space by default? Which is the sys_glbl_vrbl name?

File: src/main/java/org/folio/marccat/integration/AuthorityStorageService.java
Comment (carmentrazza) [reply src/main/java/org/folio/marccat/integration/AuthorityStorageService.java:175]: Yes, it if you always have S as indicators of the skip in filing in the authoriy tag table also for authority record

File: src/main/resources/sample-data/load-from-file/authority/name.json
Comment (carmentrazza) [reply src/main/resources/sample-data/load-from-file/authority/name.json:1]: ok

File: src/main/java/org/folio/marccat/integration/AuthorityStorageService.java
Comment (carmentrazza) [reply src/main/java/org/folio/marccat/integration/AuthorityStorageService.java:92]: Okay, it was to tell you that you have to foresee it in the future.

File: src/main/java/org/folio/marccat/dao/persistence/ItemEntity.java
Comment (carmentrazza) [reply src/main/java/org/folio/marccat/dao/persistence/ItemEntity.java:25]: In the AMICUS.S_SYS_GLBL_VRBL table the default is authority.verificationLevel = '1'.  You can change the value. The defaults were inserted with the sql 1.5 patch with the MODCAT-165.sql script

File: src/test/java/org/folio/marccat/resources/AuthorityRecordTest.java
Comment (carmentrazza) [reply src/test/java/org/folio/marccat/resources/AuthorityRecordTest.java:26]: All right

File: src/main/java/org/folio/marccat/dao/AuthorityCatalogDAO.java
Comment (carmentrazza) [reply src/main/java/org/folio/marccat/dao/AuthorityCatalogDAO.java:40]: All tags.

File: src/main/resources/sample-data/load-from-file/authority/name.json
Comment (carmentrazza) [reply src/main/resources/sample-data/load-from-file/authority/name.json:1]: Ok

File: src/main/java/org/folio/marccat/dao/AuthorityCatalogDAO.java
Comment (eturienzo) [reply src/main/java/org/folio/marccat/dao/AuthorityCatalogDAO.java]: I have done changes to do the sort

File: ramls/examples/name.json
Comment (eturienzo) [reply ramls/examples/name.json:1]: I have done the changes

File: src/main/java/org/folio/marccat/dao/persistence/ItemEntity.java
Comment (eturienzo) [reply src/main/java/org/folio/marccat/dao/persistence/ItemEntity.java]: I have done the changes. Thanks.

File: src/main/java/org/folio/marccat/integration/AuthorityStorageService.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/integration/AuthorityStorageService.java:175]: That's right, we use S in the skip in filing of authorities

**Review by carmentrazza — APPROVED** [review]

File: src/main/java/org/folio/marccat/resources/AuthorityRecordAPI.java
Comment (mreno-EBSCO) [inline src/main/java/org/folio/marccat/resources/AuthorityRecordAPI.java:43]: Could we use a different acronym here? Some people could take offense to this variable name. Perhaps don't use an acronym at all, i.e. `authorityStorageService`.

File: src/main/java/org/folio/marccat/resources/domain/AuthorityRecord.java
Comment (mreno-EBSCO) [inline src/main/java/org/folio/marccat/resources/domain/AuthorityRecord.java:15]:
```suggestion
public class AuthorityRecord extends Record {
```
There are other places where there is no space before the `{`. We should correct them all.

File: src/main/resources/database-setup/patches/MARCCAT_DB/1.6/MODCAT-210.sql
Comment (mreno-EBSCO) [inline src/main/resources/database-setup/patches/MARCCAT_DB/1.6/MODCAT-210.sql:1046]: Just curious why the commit is here vs. the end of the file?

File: ramls/schema/authority.json
Comment (LeireDiez) [reply ramls/schema/authority.json:2]: Yes, it is done

File: src/main/java/org/folio/marccat/business/cataloguing/authority/AuthorityCatalog.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/business/cataloguing/authority/AuthorityCatalog.java:82]: This values won't be changed in a production environment. They are constants that are used to configure the mappings of the tags. For example, it identifies the class of the name authority

File: ramls/schema/authority.json
Comment (LeireDiez) [reply ramls/schema/authority.json]: I have formatted with two-space, as said in https://dev.folio.org/guidelines/contributing/#consistent-whitespace

File: ramls/schema/fixedField.json
Comment (LeireDiez) [reply ramls/schema/fixedField.json]: I have formatted with two-space, as said in https://dev.folio.org/guidelines/contributing/#consistent-whitespace

File: ramls/schema/variableField.json
Comment (LeireDiez) [reply ramls/schema/variableField.json]: You're right! I have formatted with two-space, as said in https://dev.folio.org/guidelines/contributing/#consistent-whitespace

File: src/main/java/org/folio/marccat/business/cataloguing/authority/AuthorityTagImpl.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/business/cataloguing/authority/AuthorityTagImpl.java:6]: I think you are right, but this task can affect to another work team. I have created the issue https://issues.folio.org/browse/MODCAT-212 to analyze it

File: src/main/java/org/folio/marccat/business/cataloguing/authority/AuthorityCatalog.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/business/cataloguing/authority/AuthorityCatalog.java:1]: I think it should be improved in a different issue. I have created https://issues.folio.org/browse/MODCAT-213. Thank you for your recommedation!

File: src/main/java/org/folio/marccat/business/common/View.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/business/common/View.java:71]: I have made a constant for that

File: src/main/java/org/folio/marccat/config/constants/Global.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/config/constants/Global.java:185]: I think you are right, I have made changes. Let me know if you see something wrong

File: src/main/java/org/folio/marccat/dao/AuthorityCorrelationDAO.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/dao/AuthorityCorrelationDAO.java:13]: Sorry, this number has been deleted

File: src/main/java/org/folio/marccat/dao/persistence/AUT.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/dao/persistence/AUT.java]: I have changed that. encodingLevel has different values for bibliographic and for authority. Now I initialize this variable in the constructor for authority with his own value.

File: src/main/java/org/folio/marccat/dao/persistence/AuthorityCorrelationKey.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/dao/persistence/AuthorityCorrelationKey.java:7]: Done!

File: src/main/java/org/folio/marccat/dao/persistence/AuthorityHeadingTag.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/dao/persistence/AuthorityHeadingTag.java:16]: Done!

File: src/main/java/org/folio/marccat/dao/persistence/AuthorityHeadingTag.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/dao/persistence/AuthorityHeadingTag.java]: Done

File: src/main/java/org/folio/marccat/dao/persistence/AuthorityNameHeadingTag.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/dao/persistence/AuthorityNameHeadingTag.java:18]: Its constant is Global.NAME_CATEGORY. I have changed

File: src/main/java/org/folio/marccat/resources/domain/AuthorityRecord.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/resources/domain/AuthorityRecord.java:15]: I have formatted the files

File: src/main/resources/database-setup/patches/MARCCAT_DB/1.6/MODCAT-210.sql
Comment (LeireDiez) [reply src/main/resources/database-setup/patches/MARCCAT_DB/1.6/MODCAT-210.sql]: I have added at the end of the file

File: src/main/java/org/folio/marccat/business/cataloguing/authority/AuthorityCatalog.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/business/cataloguing/authority/AuthorityCatalog.java]: I have added a comment to say the reason

File: src/main/java/org/folio/marccat/business/cataloguing/authority/AuthorityHeaderFieldHelper.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/business/cataloguing/authority/AuthorityHeaderFieldHelper.java:16]: Done with the new format

File: src/main/java/org/folio/marccat/business/cataloguing/authority/AuthorityItem.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/business/cataloguing/authority/AuthorityItem.java]: Done with the new format

File: src/main/java/org/folio/marccat/config/constants/Global.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/config/constants/Global.java]: Now I have problems with duplicated lines, what do you recommend me?

File: descriptors/ModuleDescriptor-template.json
Comment (LeireDiez) [reply descriptors/ModuleDescriptor-template.json]: I think that modulePermissions is not required for authorities

File: descriptors/ModuleDescriptor-template.json
Comment (LeireDiez) [reply descriptors/ModuleDescriptor-template.json]: Done for authorities! I have created the issue https://issues.folio.org/browse/MODCAT-215 to fix the permissions of previous APIs

File: src/main/java/org/folio/marccat/business/cataloguing/authority/AuthorityCatalog.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/business/cataloguing/authority/AuthorityCatalog.java]: Done with the new format

File: src/main/java/org/folio/marccat/business/cataloguing/authority/AuthorityItem.java
Comment (LeireDiez) [reply src/main/java/org/folio/marccat/business/cataloguing/authority/AuthorityItem.java]: I have changed this function

File: ramls/schema/fixedField.json
Comment (mreno-EBSCO) [inline ramls/schema/fixedField.json:30]: Is there any structure to this field? This seems to defeat the purpose of the `additionalProperties` field. While there may be cases of fields that are just `object` in other FOLIO code, I am not aware of them. Can this field be bounded in some way or is it truly arbitrary? There are security concerns with respect to storing unstructured data.

File: src/main/java/org/folio/marccat/dao/persistence/AUT.java
Comment (mreno-EBSCO) [inline src/main/java/org/folio/marccat/dao/persistence/AUT.java:61]: Should this be updated as well? What is "librisuite" in this context?

**Review by mreno-EBSCO — APPROVED** [review]
I've added some additional things that should be considered before merging. Also, I noticed that there are 2 maven-surefire-plugin's declared in the POM file which generates a dire warning in the build. We should fix this, either with this PR on a new one.

```
[WARNING] 
[WARNING] Some problems were encountered while building the effective model for org.folio:mod-marccat:jar:2.3.0-SNAPSHOT
[WARNING] 'build.plugins.plugin.(groupId:artifactId)' must be unique but found duplicate declaration of plugin org.apache.maven.plugins:maven-surefire-plugin @ line 496, column 15
[WARNING] 
[WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
[WARNING] 
[WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
[WARNING] 
```
There are also still a lot of magic numbers that I find a bit confusing, but not enough to block this PR. We may want to consider replacing magic numbers with descriptive constants at some point.

Anyway, I am going to approve this so that when the new comments are addressed, you can just merge the PR.

**Pull request merged**
